### ISSUE TITLE
ToDo list

### ISSUE THREAD

**Issue by jarun** [jarun/nnn#1122]
Rolled from #1040.

### Cooking

- [x] add `bookmarks` directory for flexible symlinked bookmarks
- [x] new key <kbd>B</kbd> to add a symlinked bookmark for current dir
- [x] sync multiple instances of `nnn` after operation on selection
- [x] save complete per-context filter when saving sessions
- [x] disable symlink resolution for paths in `NNN_BMS` and arg `PATH`
- [x] do not end selection mode before running plugins/prompt/shell
- [x] plugin `finder` will now retain search filter history
- [x] plugin `bookmarks` removed with symlinked bookmarks support
- [x] list open locations in active contexts in help page
- [x] make option `O_MATCHFLTR` to discard filter key if no match
- [x] configurable `NNN_TMPFILE` to _cd on quit_
- [x] disable auto marking directories (use <kbd>-</kbd>)
- [x] picker mode improvements
  - open tty for input if `STDIN` is non-tty
  - truncate output file before writing
  - do not double select a file on <kbd>Enter</kbd>
- [x] no redraw during du calculation, show only current dir name
- [x] plugin `xdgdefault`: add dmenu support

### Up for grabs

- [ ] notcurses UI backend (#1014) [@KlzXS]
- [ ] open new terminal window in spawned shell dir (#759)

For anything else please discuss in this thread.

**Contribution [guideline](https://github.com/jarun/nnn/blob/master/misc/CONTRIBUTING.md).**

**Comment by lawnowner** [comment]
62a04864805535d6bc8eea4e320c505de3d59818 works well, thanks!

> The problem is using nav tools within a nav tool for navigation.

As @pvx- also pointed out, a bookmarks plugin can be handy at times!

**Comment by pvx-** [comment]
Just to repeat my older post, instead of "support applying filter by key (like bookmarks/plugins)" I suggest "allow to set filter from plugin using NNN_PIPE", if it is technically possible; I tried to look at the code myself, but the filter stuff is a bit too complicated for me ;).

Also if you have a symlink as directory bookmark, it is followed when used (symlink target is set as current directory); this does not happen when setting directory from plugin. Again I tried to look at the code, but couldn't find where it happens.

**Comment by jarun** [comment]
> but the filter stuff is a bit too complicated for me

Yes, even I don't want to deal with it. normal string/regular/case/ignore case... too many combos.

> a symlink as directory bookmark

In NNN_BMS?

> this does not happen when setting directory from plugin

You mean if you pass a symlink from plugin, it is not followed?

**Comment by jarun** [comment]
`NNN_BMS` is resolved using `realpath()` which resolves symlinks. The paths from plugins aren't. You can modify the plugin you use to achieve the same behaviour. A central place would be `.nnn-plugin-helper` for plugins those use it.

**Comment by pvx-** [comment]
Yes `NNN_BMS`; what I want is the exact opposite :), so that it is *not* resolved when set from bookmark (if you want that, just put the target directory directly). I use a few symlinks this way and would really prefer to have them set as current directory (instead of their target), if possible.

**Comment by lawnowner** [comment]
> instead of "support applying filter by key (like bookmarks/plugins)" I suggest "allow to set filter from plugin using NNN_PIPE"

I've never needed to apply a filter via a plugin, although a workaround can be: http://github.com/jarun/nnn/tree/master/plugins#quick-find-using-fd

**Comment by jarun** [comment]
@pvx- can you try commit da193a4d3fef1a26fc51a4c421398f5ce8337ff0.

Also, there are some more uses of `realpath()` in the code. Can you check them and see which are really necessary? We can replace the rest with `abspath()`.

**Comment by jarun** [comment]
I think @lawnowner's solution is a good way for users to have plugins of their own defined by plugins.

**Comment by pvx-** [comment]
> @pvx- can you try commit [da193a4](https://github.com/jarun/nnn/commit/da193a4d3fef1a26fc51a4c421398f5ce8337ff0).

Thanks, works great.

> Also, there are some more uses of `realpath()` in the code. Can you check them and see which are really necessary? We can replace the rest with `abspath()`.

Not sure I'm the best person to check :), but I tried. One that probably can be changed is `initpath` (to leave initial path as-is even if it's a symlink), also in picker option (`-p`) (if user selects symlink it should return it, not its target); the one in handle_archive may be needed, in SEL_NAV_IN as well (if I see correctly it just checks the symlink target is accessible).

**Comment by pvx-** [comment]
> > instead of "support applying filter by key (like bookmarks/plugins)" I suggest "allow to set filter from plugin using NNN_PIPE"
> 
> I've never needed to apply a filter via a plugin, although a workaround can be: [http://github.com/jarun/nnn/tree/master/plugins#quick-find-using-fd](https://github.com/jarun/nnn/tree/master/plugins#quick-find-using-fd)

That is what @jarun suggested the first time :), but it's not usable for what I need (like select those files and move them). But if I'm the only one with such weird workflow, I'll manage with pasting the regex into filter input, it's not like I need it that often :).

I just think that if someone needs bookmarks for filters, doing it with a plugin is much more flexible, you can add history, etc. (I got really used to having history in every input in mc ;). But it seems nobody really needs it enough to implement it, so that's that ;).

**Comment by jarun** [comment]
> but it's not usable for what I need (like select those files and move them).

Why not? You can select, move, do whatever you want...

**Comment by jarun** [comment]
You can have your often used list of filters in the plugin, select the one you want and the listing will be done in a separate context. You can select those files and run all operations on them.

**Comment by pvx-** [comment]
Thanks, I will try to experiment with list mode more, there will be probably a way to use them to do what I need. When you do something a certain way for many years it's not easy to look at it "fresh" again ;).

**Comment by lawnowner** [comment]
Your idea of bookmarks for filters isn't different than bookmarks for directories/files when both are implemented with a plugin, except the plugin uses the opcode `c` to go to a bookmarked directory/file whereas it uses the opcode `l` to filter and then list a number of paths. The list mode (http://github.com/jarun/nnn/wiki/concepts#find-and-list) gives a temporary view with symlinks since it allows entries to be from different directories, and possibly also applying different filters to different directories, which is already better than having `NNN_FLTRS` (or mc for that matter) I think, that is assuming you really need to apply filters with a key ;] Any operation is performed on the targets of symlinks, including running plugins, although the temporary view isn't updated, e.g. deleting any symlink deletes the target but the symlink remains in the list.

>`initpath` (to leave initial path as-is even if it's a symlink)

:+1:

**Comment by jarun** [comment]
@pvx- once you are done with the filter thing, can you please document it in the Wiki "How to" page?

I have removed symlink resolution in initpath in commit 59d90aa2ce9fb6ed4a12212fd61fc428d5f277e0.

**Comment by jarun** [comment]
> in picker option (-p) (if user selects symlink it should return it, not its target)

Can you point me to the code?

**Comment by pvx-** [comment]
> @pvx- once you are done with the filter thing, can you please document it in the Wiki "How to" page?

What exactly should I document?

> I have removed symlink resolution in initpath in commit [59d90aa](https://github.com/jarun/nnn/commit/59d90aa2ce9fb6ed4a12212fd61fc428d5f277e0).

Hm, no realpath() in code anymore, but if I open `nnn` in a symlinked directory, current path is resolved (like if `/tmp/sym` is a symlink to `/tmp/real`, opening `nnn` in `/tmp/sym/test` ends up with `/tmp/real/test` as current directory).

> > in picker option (-p) (if user selects symlink it should return it, not its target)
> 
> Can you point me to the code?

You already did it in commit above (`selpath = `).

**Comment by raj-magesh** [comment]
Are there are any plans to support selecting multiple files using click-and-drag for mouse users? If not, is it technically possible to implement that as a plugin?

PS: Thanks for the awesome file manager :)

**Comment by luukvbaal** [comment]
Not sure if click-and-drag is feasible but what you can do is configure `NNN_MCLICK` to range select i.e. `NNN_MCLICK='m' nnn` and use middle mouse button to mark the begin and and of your range selection.

**Comment by 0xACE** [comment]
> Are there are any plans to support selecting multiple files using click-and-drag for mouse users? If not, is it technically possible to implement that as a plugin?

I experimented with this, but honestly it felt too much like:

![xkcd-927-standards_2x](https://user-images.githubusercontent.com/1482942/128878101-1c31ea6e-fb7b-4d08-9f90-2a145219ad2b.png)

and felt that it wasn't worth the effort. KDE's `Konsole` seems like a candidate to implement mouse drag and drop support and what not, sometime in the future. So don't expect any magic unless your terminal is magical...

A very rare usecase for me is when I do complex markings i usually just select the entire folder and edit the selection buffer in `vim` with my own complex filters using macros (`q`), `:g` or `:v`.

**Comment by jarun** [comment]
@pvx- 

> What exactly should I document?

Instead of documenting, I think we can make it a plugin if you can come up with such a script.

> opening nnn in /tmp/sym/test ends up with /tmp/real/test

Fixed that at commit 97ac88dec2fe4cd2bde49d50817e359531c7280c.

@raj-magesh 

@luukvbaal's solution is the closest you get. Technically, I don't think we can support it everywhere anyway. For example, termux doesn't support gestures as far as I know.

**Comment by jarun** [comment]
@pvx- 

the workflow I was thinking for the plugin is - have and array of predefined filters, list and select by number, apply it through find. You can limit it to the current dir by setting the depth if you want.

**Comment by pvx-** [comment]
I was quite busy today, but I put together a first shot, any comments are welcome:

```sh
findfn="/tmp/.nnn/find_bkm"

cat -n "$findfn"

if type rlwrap >/dev/null 2>&1; then
    fexpr=$(rlwrap -s -1 -H "$findfn" sh -c 'read -r -p "Expression or number: " REPLY && echo $REPLY')
else
    read -p "Expression or number: " fexpr
fi

# for number input read from bookmark file
if [ "$fexpr" -eq "$fexpr" ] 2> /dev/null
then
    fexpr=`sed -n -e "$fexpr"p "$findfn"`
fi

if [ -n "$fexpr" ]; then
    printf "%s" "+l" > "$NNN_PIPE"
    eval "find ./ -maxdepth 1 -type f $fexpr -print0" > "$NNN_PIPE"
fi
```

* `findfn` is bookmarks filename, lines with find expressions like `-name '*.txt'` (put it into nnn config dir or is there some other place for it?)
* I thought about which args to put as default, decided on maxdepth and type; if you think less/more is better let me know (also every user can change it, I personally don't use a single plugin as-is, only as a base for my own)
* maybe some input-checking and error-reporting would be useful, but I don't know how much is reasonable :)

**Comment by jarun** [comment]
> pasted it

what is paste? cp/mv?

Anyway, are these 2 instances? In that case I don't think there is any simple way to left left instance of `nnn` know that you have used the selection. I don't want to go polling or IPC. Why can't you do this in 2 contexts?

**Comment by jarun** [comment]
@pvx- 

> findfn is bookmarks filename

OK. So you were looking for a way to find bookmarks easily. I am not sure if any of the existing plugins do that already.
Anyway, I can't say much about scripts. Please get help from @luukvbaal or @KlzXS or @Kabouik.

I was under the impression you were looking for a way to save some predefined filters and apply them.

**Comment by jarun** [comment]
@N-R-K the issue is fixed at commit 06e267642141c74e1d2d63d94b343160029bc717.

**Comment by pvx-** [comment]
> @pvx-
> 
> > findfn is bookmarks filename
> 
> OK. So you were looking for a way to find bookmarks easily. I am not sure if any of the existing plugins do that already.
> Anyway, I can't say much about scripts. Please get help from @luukvbaal or @KlzXS or @Kabouik.
> 
> I was under the impression you were looking for a way to save some predefined filters and apply them.

Sorry for the confusion, by "bookmarks" here I mean predefined filters ("find" command expressions to be exact). And as I wrote repeatedly before, I only reacted to the "support applying filter by key (like bookmarks/plugins)" item already present in your todo list; I'm not sure it would useful for enough people to be included in plugins, but if you think it is, I'll be glad to help; let me know the best way where to discuss it (new issue/dev forum/...?).

What I personally need is even more niche, specifically to edit the expressions slightly before applying, that's why I use rlwrap :).

**Comment by jarun** [comment]
We won't have rlwrap as a dep so you can have it in your local copy.

What's the output of: `cat -n "$findfn"`?

**Comment by N-R-K** [comment]
> Why can't you do this in 2 contexts?

That's what I typically do but I noticed this while trying to experiement with "dual plane" mode via tmux. Anyhow, https://github.com/jarun/nnn/commit/06e267642141c74e1d2d63d94b343160029bc717 fixed the issue on ``mv``, but it still persists on ``cp``.

**Comment by jarun** [comment]
When you say `cp`, what do did you do? Did you do a selection operation or did you just `cp` the file from one place to another? Note that you have to operate on the selection.

**Comment by jarun** [comment]
OK. I see what you mean. You have to understand that in this case the file is not moved so there is no change in the directory in the instance where you selected the file. So are doing this on directory notification (read _change_), so  mv and rm would work. I don't think it's a big deal anyway as long as we show the selection info in the bar and the symbols.

I will add this to the docs.

**Comment by N-R-K** [comment]
> When you say `cp`, what do did you do?

Pressed <kbd>p</kbd> on the other window.

> as long as we show the selection info in the bar and the symbols.

The selection info on the bar and symbol visually shows that there's an active selection when there's none (since the other nnn already copied the file and cleared selection). And trying to do any operation on that selection fails with the error I described above.

**Comment by jarun** [comment]
Commit 42061b0399008f0ae7a2863e8a8e0e0465c41d62 detects this and blocks the operation early with a message.

**Comment by pvx-** [comment]
> We won't have rlwrap as a dep so you can have it in your local copy.

Of course, it's optional in the script (used only if present).

> What's the output of: `cat -n "$findfn"`?

Adds number to each line.

> > put it into nnn config dir or is there some other place for it?
> 
> I would say put it in the plugin file and use as an array. It's OK to have bash for this.

Usually people want it separated :) and rlwrap needs it (it's given as history file for readline); but if you prefer it in the file, I can remove it. I use bash a lot but I see that nnn is used on many platforms, so I used string instead (hopefully more compatible):

```sh
# list of preset expressions to choose from (one per line)
read -r -d '' fpreset <<EOF
-name '*.mkv'
-size +10M
-name '*.pdf' -mtime -1
EOF

echo "$fpreset" | cat -n

read -p "Expression or number: " fexpr

if [ "$fexpr" -eq "$fexpr" ] 2> /dev/null
then
    fexpr=`echo "$fpreset" | sed -n -e "$fexpr"p`
fi

if [ -n "$fexpr" ]; then
    printf "%s" "+l" > "$NNN_PIPE"
    eval "find ./ -maxdepth 1 -type f $fexpr -print0" > "$NNN_PIPE"
fi
```

**Comment by jarun** [comment]
@luukvbaal @KlzXS can you please review?

**Comment by KlzXS** [comment]
I will take a look. Should we move the conversation somewhere else?
I wasn't keeping an eye on the TODO list and now there seem to be at least two conversations going on in parallel.

**Comment by jarun** [comment]
Yes we can resume in the dev room.

**Comment by wazabees** [comment]
I've been a (very) long time user of Total Commander on Windows, and later Double Commander on Linux. I'm looking to use nnn instead, and as such I'm wondering if a feature I frequently used is possible in nnn? In both Total Commander and Double Commander I can use the keyboard shortcut `Ctrl+B` to get a list of all files in subdirectories from my current directory. The shortcut is described as "**Directory branch: Show contents of current dir and all subdirs in one list"**.

This is useful for me, if I have a whole bunch of subdirectories and I want to move the files contained in them into one directory. The result would be similar to running the command `find . -mindepth 2 -type f -exec mv -i -- {} . \;`, but with the possibility to choose the files I want.

I'm not sure I explained this one very well. :)

**Comment by jarun** [comment]
2 things:

- Your current file managers support this because they don't allow selecting file from multiple directories. You can select files from multiple directories in `nnn`.
- You can use plugin `finder` to run the command `find . -mindepth 2 -type f -print0`, list the matches, select them by navigating and them move to another location using `^V`.

**Comment by N-R-K** [comment]
Would it be an acceptable feature request to ask for adding a new bookmark implementation that is managed via symlinks? This can be a runtime or a compile time toggle so that it doesn't break the existing workflow.

The idea here is very simple, when <kbd>b</kbd> is pressed, go inside `$NNN_BOOKMARK_DIR`. The [bookmarks](https://github.com/jarun/nnn/blob/master/plugins/bookmarks) plugin already does something similar, but it takes 1 extra keypress and depends on an external menu (fzf). There can also be another variable, `$NNN_BFILTER` which controls what to do after going inside the bookmark dir, if it's set to 1 then activate filter mode, type-to-nav mode if 2 and do nothing if 0 or unset.

The rational behind it is that it can be synced effortlessly to multiple applications. You can point, for example your ``$CDPATH`` , filepicker or any other app to this single unified directory. Another additional benefit is that you can create/delete/manage bookmarks right from `nnn` without having to edit environment variables.

I have been playing around with this for the past couple days and it integrates very well with the environment and overall feels much more seamless compared to using a plugin.

**Comment by 0xACE** [comment]
I don't understand your request, but if you can't have it, this simple patch ought to help: You can use this to trigger plugins with 1 key. Caveat is that the key registered in `nnn` will have to be the same plugin used by `NNN_PLUG`. in this patch `NNN_PLUG=Z:foo.sh` should launch the plugin `foo.sh`. If you want it to link to another key, you'll have to either introduce a new array or do a little more invasive patch which wont be fun to maintain...
```
diff --git a/src/nnn.c b/src/nnn.c
index fe5d650..e575bee 100644
--- a/src/nnn.c
+++ b/src/nnn.c
@@ -3095,7 +3114,11 @@ try_quit:

        for (i = 0; i < (int)ELEMENTS(bindings); ++i)
                if (c == bindings[i].sym)
+               {
+                       if (bindings[i].act == SEL_FAST_PLUGIN)
+                               ungetch(c);
                        return bindings[i].act;
+               }

        return 0;
 }
@@ -7442,6 +7465,8 @@ nochange:

                        goto begin;
                }
+               case SEL_FAST_PLUGIN:
+                       pkey = getch(); // fallthrough
                case SEL_PLUGIN:
                        /* Check if directory is accessible */
                        if (!xdiraccess(plgpath)) {
diff --git a/src/nnn.h b/src/nnn.h
index 6e934df..5ddefc7 100644
--- a/src/nnn.h
+++ b/src/nnn.h
@@ -101,6 +101,7 @@ enum action {
        SEL_HELP,
        SEL_AUTONEXT,
        SEL_EDIT,
+       SEL_FAST_PLUGIN,
        SEL_PLUGIN,
        SEL_SHELL,
        SEL_LAUNCH,
@@ -250,6 +251,7 @@ static struct key bindings[] = {
        { 'e',            SEL_EDIT },
        /* Run a plugin */
        { ';',            SEL_PLUGIN },
+       { 'Z',            SEL_FAST_PLUGIN },
        /* Run command */
        { '!',            SEL_SHELL },
        { CONTROL(']'),   SEL_SHELL },
```
I haven't tested it, may doom guy be on your side ![doomguy](https://user-images.githubusercontent.com/1482942/129496185-346692eb-e68b-4290-9380-a400e0f35d4f.png)

**Comment by N-R-K** [comment]
Thanks for the patch, however I already have what I want on my local build. It doesn't preserve backwards compatibility and has hard-coded values but maybe you can understand what I mean better by looking at the patch. I wasn't happy having to use an external tool when nnn is perfectly capable of visiting a given directory and activating filter mode (or in this case type-to-nav).

```diff
diff --git a/src/nnn.c b/src/nnn.c
index fe5d650..a3789a2 100644
--- a/src/nnn.c
+++ b/src/nnn.c
@@ -434,6 +434,7 @@ static char *enveditor;
 static char *pager;
 static char *shell;
 static char *home;
+static char *bookmark_dir;
 static char *initpath;
 static char *cfgpath;
 static char *selpath;
@@ -619,6 +620,7 @@ static char * const utils[] = {
 #define MSG_INVALID_KEY  40
 #define MSG_NOCHANGE     41
 #define MSG_DIR_CHANGED  42
+#define MSG_BKDIR_NOTSET 43
 
 static const char * const messages[] = {
 	"",
@@ -664,6 +666,7 @@ static const char * const messages[] = {
 	"invalid key",
 	"unchanged",
 	"dir changed, range sel off",
+	"NNN_BOOKMARK_DIR not set"
 };
 
 /* Supported configuration environment variables */
@@ -6365,7 +6368,7 @@ static bool browse(char *ipath, const char *session, int pkey)
 	     runfile[NAME_MAX + 1] __attribute__ ((aligned));
 	char *path, *lastdir, *lastname, *dir, *tmp;
 	pEntry pent;
-	enum action sel;
+	enum action sel = SEL_NULL;
 	struct stat sb;
 	int r = -1, presel, selstartid = 0, selendid = 0;
 	const uchar_t opener_flags = (cfg.cliopener ? F_CLI : (F_NOTRACE | F_NOSTDIN | F_NOWAIT));
@@ -6518,7 +6521,7 @@ nochange:
 		if (!isatty(STDIN_FILENO) && !g_state.picker)
 			return EXIT_FAILURE;
 
-		sel = nextsel(presel);
+		sel = (sel == SEL_BOOKMARK) ? SEL_MFLTR : nextsel(presel);
 		if (presel)
 			presel = 0;
 
@@ -6856,13 +6859,29 @@ nochange:
 		case SEL_CDBEGIN: // fallthrough
 		case SEL_CDLAST: // fallthrough
 		case SEL_CDROOT:
-			dir = (sel == SEL_CDHOME) ? home
-				: ((sel == SEL_CDBEGIN) ? ipath
-				: ((sel == SEL_CDLAST) ? lastdir
-				: "/" /* SEL_CDROOT */));
+		case SEL_BOOKMARK:
+			switch (sel) {
+				case SEL_CDHOME:
+					dir = home; break;
+				case SEL_CDBEGIN:
+					dir = ipath; break;
+				case SEL_CDLAST:
+					dir = lastdir; break;
+				case SEL_BOOKMARK:
+					dir = bookmark_dir;
+					cfg.filtermode = 0;
+					break;
+				default: // CDROOT
+					dir = "/"; break;
+			}
 
 			if (!dir || !*dir) {
-				printwait(messages[MSG_NOT_SET], &presel);
+				if (sel == SEL_BOOKMARK) {
+					printwait(messages[MSG_BKDIR_NOTSET], &presel);
+					sel = SEL_NULL;
+				} else {
+					printwait(messages[MSG_NOT_SET], &presel);
+				}
 				goto nochange;
 			}
 
@@ -6882,17 +6901,6 @@ nochange:
 
 			/* SEL_CDLAST: dir pointing to lastdir */
 			xstrsncpy(newpath, dir, PATH_MAX); // fallthrough
-		case SEL_BOOKMARK:
-			if (sel == SEL_BOOKMARK) {
-				r = (int)handle_bookmark(mark, newpath);
-				if (r) {
-					printwait(messages[r], &presel);
-					goto nochange;
-				}
-
-				if (strcmp(path, newpath) == 0)
-					break;
-			}
 
 			/* In list mode, retain the last file name to highlight it, if possible */
 			cdprep(lastdir, listpath && sel == SEL_CDLAST ? NULL : lastname, path, newpath)
@@ -8284,6 +8292,9 @@ int main(int argc, char *argv[])
 	if (!setup_config())
 		return EXIT_FAILURE;
 
+	/* Get bookmark directory */
+	bookmark_dir = getenv("NNN_BOOKMARK_DIR");
+
 	/* Get custom opener, if set */
 	opener = xgetenv(env_cfg[NNN_OPENER], utils[UTIL_OPENER]);
 	DPRINTF_S(opener);
diff --git a/src/nnn.h b/src/nnn.h
index 6e934df..40d9fa7 100644
--- a/src/nnn.h
+++ b/src/nnn.h
@@ -44,7 +44,8 @@
 
 /* Supported actions */
 enum action {
-	SEL_BACK = 1,
+	SEL_NULL,
+	SEL_BACK,
 	SEL_OPEN,
 	SEL_NAV_IN,
 	SEL_NEXT,
```

**Comment by 0xACE** [comment]
> Thanks for the patch, however I already have what I want on my local build.

Haha, oddly enough I understood the patch better than the text.

These threads get out of hand frequently, and i practically ignore any scripting topics, you'll stick out more in a PR...

If the patch works: just submit a PR. It's cleaner that way... I'm not saying it will get adopted. But a PR is a better channel than discussing hypotheticals all day, specially in this case where you have a working prototype, cheers & good luck :)

**Comment by jarun** [comment]
@N-R-K what happens if someone wants to cancel the operation? i.e. she goes into the bookmarks directory but wants to cancel and return back to the original directory where she was?

**Comment by jarun** [comment]
Why does the plugin take an extra keypress and why can't we just have a simple list to pick from instead of fzf?

**Comment by jarun** [comment]
BTW, note that we have a way to mark directories temporarily.

**Comment by N-R-K** [comment]
> what happens if someone wants to cancel the operation?

In normal mode and type-to-nav mode, simply pressing <kbd>-</kbd> will do the trick. In filter mode since <kbd>-</kbd> doesn't have special meaning, an extra <kbd>esc</kbd> will need to be pressed before.



> Why does the plugin take an extra keypress

Because plugins need to be activated with <kbd>;</kbd> first.

> why can't we just have a simple list to pick from instead of fzf?

I don't quite understand this. Do you mean the default nnn bookmarks? It _works_ but that means managing my nnn bookmarks separately. With symlinks, I can point my filepicker, shell as well as nnn into one directory and no need to manage separate bookmarks.

Also is there a way for a plugin to activate filter/type-to-nav mode? I did not see any [opcode](https://github.com/jarun/nnn/tree/master/plugins#send-data-to-nnn) for that. If a plugin can activate filter/type-to-nav mode, then I can probably do what I'm trying to do with a plugin instead.

**Comment by jarun** [comment]
I think we can have the following workflow:

Have a new directory `bookmarks` within config dir for bookmarks. We still retain the current implementation. if the user presses `Enter` at the bookmark prompt we take him to the bookmarks dir as we do with plugins.

> Because plugins need to be activated with ; first.

You would still need a key to go to bookmark dir.

> Also is there a way for a plugin to activate filter/type-to-nav mode?

No. Pressing `/` or `^N` is not so difficult.

**Comment by jarun** [comment]
I have a working implementation.

I am not adding any code to create new bookmarks. To add bookmarks, user can select the directory to mark, go to the `bookmarks` directory, press `n` and create the symlink.

**Comment by jarun** [comment]
Also, if someone doesn't set `NNN_BMS` and there is no marked directory, we will directly take her to the `bookmarks` directory.

With this, I will remove the `bookmarks` plugin as well.

**Comment by N-R-K** [comment]
> To add bookmarks, user can select the directory to mark, go to the `bookmarks` directory, press `n` and create the symlink.

Yes that was my intention as well. Also I think the bookmark directory should be configurable via an env var for flexibility. If it's not set then falling back to a default seems reasonable.

**Comment by jarun** [comment]
> Also I think the bookmark directory should be configurable via an env var for flexibility

I don't think so. We can't keep adding env vars.

**Comment by N-R-K** [comment]
Hmm, fair enough. Can this bookmark dir be a symlink to someplace else? If so then I guess env var is not needed.

**Comment by jarun** [comment]
Why?

**Comment by jarun** [comment]
You can always create symlinks to this bookmark dir instead.

**Comment by N-R-K** [comment]
For better flexibility. The plugin dir on my `.config/nnn` is a symlink for example. Whenever I do a git pull the plugin dir, being a symlink, gets updated automatically.
But it's not that big of an issue since I can either patch it to my liking locally or do what you suggested and symlink to nnn bookmark dir instead.

**Comment by jarun** [comment]
I think you can make `bookmarks` a symlink as well. try it.

I would leave it at that. Don't want to litter everything around.

> Whenever I do a git pull the plugin dir, being a symlink, gets updated automatically.

That's unsupported and dangerous. We do not guarantee backward compatibility. These are things I like to avoid. Regular users mess things up with too much flexibility and come back asking us to fix their local problems.

**Comment by N-R-K** [comment]
> I think you can make `bookmarks` a symlink as well. Try it.

Seems to be working as expected.

**Comment by N-R-K** [comment]
@jarun https://github.com/jarun/nnn/commit/86648ab391483cb72308a5035c55d6ca55b01254 is creating `bookmarks`, `plugins`, `mounts`, `sessions` directories inside `~/.config/` instead of `~/.config/nnn/`. I hope this is not intended.

**Comment by lawnowner** [comment]
@N-R-K I don't have any of those directories in my `~/.config`. However, let me mention this: https://github.com/jarun/nnn/commit/46294e9f3d613fb56d381e1637af6c0d7c30e9e4#commitcomment-54979648

**Comment by N-R-K** [comment]
@lawnowner  `bookmarks` or `plugins` being a symlink triggers this for me.

**Comment by 0xACE** [comment]
> [86648ab](https://github.com/jarun/nnn/commit/86648ab391483cb72308a5035c55d6ca55b01254) is creating `bookmarks`, `plugins`, `mounts`, `sessions` directories inside `~/.config/` instead of `~/.config/nnn/`. I hope this is not intended.

I'm not familiar with w/e these `bookmark` things are but:
```
make debug # not necessary if you have debug symbols in your nnn binary
strace -ke mkdir -o /tmp/mkdir.log ./nnn
```
Share `/tmp/mkdir.log` and maybe what you did... (maybe we debug print `SEL_*` ?)

gl

**Comment by N-R-K** [comment]
> I'm not familiar with w/e these `bookmark` things are but:

It's not just `bookmarks`, `plugins` being a symlink also triggers this.
This specific line seems to be responsible for this and changing it from 13 to 14 fixes the issue.
```diff
@@ -7988,7 +7988,7 @@ static bool setup_config(void)
 
 	if (xdg) {
 		xstrsncpy(cfgpath, xdgcfg, len);
-		r = len - 13; /* subtract length of "/nnn/sessions" */
+		r = len - 14; /* subtract length of "/nnn/sessions" */
 	} else {
 		r = xstrsncpy(cfgpath, home, len);
 
```

Anyhow, https://github.com/jarun/nnn/commit/9d4330e382e5b3542bd5f51690c8bb0bdbff1fd8 fixed it for me as well and is a much cleaner approach. Thank you @0xACE .

**Comment by jarun** [comment]
At commit 78ea6702f7dae427c7a102cb8420533b7ebcc6ae it's possible to add bookmarks on the fly with `B`.
